experiments/compositions/expanders/cel-expander/pkg/resource/resource.go

```diff
@@ -0,0 +1,98 @@
+package resource
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/google/cel-go/common/types"
+	"gopkg.in/yaml.v2"
+)
+
+type Resource struct {
+	Name      string
+	Data      map[string]interface{}
+	Raw       []byte
+	Variables []*Variable
+}
+
+func NewResourceFromRaw(name string, raw []byte) (*Resource, error) {
+	var data map[string]interface{}
+	err := yaml.Unmarshal(raw, &data)
+	if err != nil {
+		return nil, err
+	}
+
+	variables, err := extractVariables(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	resource := &Resource{
+		Name:      name,
+		Data:      data,
+		Raw:       raw,
+		Variables: variables,
+	}
+	return resource, nil
+}
+
+func (r *Resource) replaceVariables(vars map[string]string) []byte {
+	copy := bytes.Clone(r.Raw)
+	for expr, elem := range vars {
+		trapRegex := regexp.MustCompile(regexExpression(expr))
+		copy = trapRegex.ReplaceAll(copy, []byte(elem))
+	}
+	return copy
+}
+
+func (r *Resource) ApplyResolvedVariables() error {
+	vars := make(map[string]string)
+	for _, variable := range r.Variables {
+		if variable.ResolvedValue != nil {
+			switch v := variable.ResolvedValue.(type) {
+			case string:
+				vars[variable.Expression] = v
+			case types.String:
+				vars[variable.Expression] = v.Value().(string)
+			case int, int64, int32, int16, int8:
+				vars[variable.Expression] = fmt.Sprintf("%d", v)
+			case types.Int:
+				vars[variable.Expression] = fmt.Sprintf("%d", v.Value().(int))
+			case float64, float32:
+				vars[variable.Expression] = fmt.Sprintf("%f", v)
+			case types.Double:
+				vars[variable.Expression] = fmt.Sprintf("%f", v.Value().(float64))
+			case bool:
+				vars[variable.Expression] = fmt.Sprintf("%t", v)
+			case types.Bool:
+				vars[variable.Expression] = fmt.Sprintf("%t", v.Value().(bool))
+			case types.Null:
+				vars[variable.Expression] = ""
+			default:
+				return fmt.Errorf("unknown variable type: %v: type=%v", variable.ResolvedValue, v)
```

Comment:
I may be wrong but I thought there's some function like fmt.Sprintf("%#v", v) that just convert any type to string?